document Hegg Dongle (#642)
esphome/esphome-devices: src/docs/devices/Hegg-Dongle/index.md

|        |                                                   | P1 data request does not appear to be exposed                      |
|        | `5V_in`                                           | tied to USB power and the P1 power pin (pin 1)                     |
| RST    | `rst`                                             | reboots the ESP when briefly tied to ground                        |

## Flashing procedure

The text below assumes you removed the plastic wrap around the device.
A (so far) untested alternative is to poke a few pins through the plastic at the bottom to reach the pads.
If you try this, please update this document with your experiences!

Using the pads, or just the pins on the ESP12F, flashing this device is a pretty "normal" procedure, except that GPIO3/RXD0 is also used to receive data from the smart meter.
The DSMR/P1 smart meter protocol uses inverted logic levels for serial, and the dongle contains an inverter circuit to turn these back around to something the ESP can work with.
It turns out that this inverter circuit makes RXD0 unusable if the physical Data pin on the P1 port is left floating.

To flash:

> * Make RXD0 idle by connecting `data_in` to Ground, or tie the corresponding pins (5 and 6, the topmost two) on the P1 connector end together.

Reviewer: ```suggestion
* Make GPIO3/RXD0 idle by connecting `data_in` to GND, or tie the corresponding pins (5 and 6, the topmost two) on the P1 connector end together.
```

> * Connect either the `txd` pad or the TXD0 pin to RX on your serial converter

Reviewer: ```suggestion
* Connect either the `txd` pad or the GPIO1/TXD0 pin to RX on your serial converter
```

> * Connect either the `rxd`, or `data_in` pad, or pin 5 of the P1 socket, or the RXD0 pin to TX on your serial converter. If you use `data_in` or the P1 pin, you likely need to invert your signal.

Reviewer: ```suggestion
* Connect either the `rxd`, or `data_in` pad, or pin 5 of the P1 socket, or the GPIO3/RXD0 pin to TX on your serial converter. If you use `data_in` or the P1 pin, you likely need to invert your signal.
```

> * Power by connecting GND and Power to your serial converter, or plugging in USB. Presumably you can also power the dongle via 5V+GND on the P1 connector. Have GND connected to GPIO0 during powerup, or briefly tie RST to GND with GPIO0 also tied to GND to reset into flashing mode.

Reviewer: ```suggestion
* Power by connecting GND and VCC to your serial converter, or plugging in USB. Presumably you can also power the dongle via 5V+GND on the P1 connector. Have GND connected to GPIO0/FLASH during powerup, or briefly tie RST to GND with GPIO0/FLASH also tied to GND to reset into flashing mode.
```

---
title: Hegg Dongle / EARN-E energiemonitor
date-published: 2024-01-28
type: misc
standard: eu
board: esp8266
---

## Introduction

FIXME: link to a few product pages here

## Product Images

FIXME these photos are terrible and also somehow became PNGs

FIXME add wrapped photo

![full (unwrapped) device](heggfull.png)
![PCB front](heggfront.png)
![PCB back](heggback.png)

## GPIO Pinout

![pinout scratch note](heggpins.jpg)

All "pads" mentioned in this table are on the back of the board.

red/green/blue are for the RGB LED right next to the ESP.

> | Pin    | Function                     |
> | ------ | ---------------------------- |
> | GND    | many; connected to one side of the (unpopulated) SW1 header; also available as the unlabeled pad |
> | GPIO0  | FLASH, connected to the other side of the (unpopulated) SW1 header
> | GPIO1  | TXD0, "txd" pad |
> | GPIO3  | RXD0, "rxd" pad, "data_in" pad via inverter |
> | GPIO12 | green 
> | GPIO13 | red
> | GPIO14 | blue
> | VCC    | power, connected to "3.3v" pad
> | ???    | P1 data request?

Reviewer: ```suggestion
| Pin    | Function                     |Pad         |
| ------ | ---------------------------- | --------------------------- |
| GND    | many; connected to one side of the (unpopulated) SW1 header; also available as the unlabeled pad | Unlabeled |
| GPIO0  | FLASH, connected to the other side of the (unpopulated) SW1 header||
| GPIO1  | TXD0| txd |
| GPIO3  | RXD0 | RXD or data_in via inverter|
| GPIO12 | green|| 
| GPIO13 | red||
| GPIO14 | blue||
| VCC    | power|3.3v|
| ???    | P1 data request?||
```

## Flashing procedure

> Using the pads, or just the pins on the ESP12F, flashing this device is a pretty "normal" procedure, except that GPIO3/RXD0 is also used to receive data from the smart meter.
> The DSMR/P1 smart meter protocol uses inverted logic levels for serial, and the dongle contains an inverter circuit to turn these back around to something the ESP can work with.
> It turns out that this inverter circuit makes RXD0 unusable if the physical Data pin on the P1 port is left floating.
> To make RXD0 idle, it is enough to tie P1 Data to Ground.
> To do that, you can either use the pads (`data_in` and FIXME is there a GND pad? perhaps next to 3.3v?), or tie them together on the P1 connector end.

Reviewer: ```suggestion
Using the pads, or just the pins on the ESP12F, flashing this device is a pretty "normal" procedure, except that GPIO3/RXD0 is also used to receive data from the smart meter.
The DSMR/P1 smart meter protocol uses inverted logic levels for serial, and the dongle contains an inverter circuit to turn these back around to something the ESP can work with.
It turns out that this inverter circuit makes RXD0 unusable if the physical Data pin on the P1 port is left floating.

To flash:
* Make RXD0 idle by connecting P1 Data to Ground, or tie the corresponding pins on the P1 connector end together.
* Connect either the txd pad or the TXD0 pin to RX on your serial converter
* Connect either the rxd or data_in pad or the RXD0 pin to TX on your serial converter
* Power by connecting GND and Power to your serial converter
* Flash as normal (FIXME: do we need to add speeds / serial settings?)

```
